bundle/config/mutator/python/apply_python_mutator.go

```diff
@@ -0,0 +1,190 @@
+			return dyn.InvalidValue, err
+		}
+
+		return merge.Override(leftRoot, rightRoot, createOverrideVisitor(ctx, m.phase))
+	})
+
+	return diag.FromErr(err)
+}
+
+func (m *applyPythonMutator) runPythonMutator(ctx context.Context, rootPath string, pythonPath string, root dyn.Value) (dyn.Value, error) {
+	args := []string{
+		pythonPath,
+		"-m",
+		"databricks.bundles.build",
+		"--phase",
+		string(m.phase),
+	}
+
+	// we need to marshal dyn.Value instead of bundle.Config to JSON to support
+	// non-string fields assigned with bundle variables
+	rootConfigJson, err := json.Marshal(root.AsAny())
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to marshal root config: %w", err)
+	}
+
+	logWriter := newLogWriter(ctx, "stderr: ")
+
+	stdout, err := process.Background(
+		ctx,
+		args,
+		process.WithDir(rootPath),
+		process.WithStderrWriter(logWriter),
+		process.WithStdinReader(bytes.NewBuffer(rootConfigJson)),
+	)
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("python mutator process failed: %w", err)
+	}
+
+	// we need absolute path, or because later parts of pipeline assume all paths are absolute
+	// and this file will be used as location
+	virtualPath, err := filepath.Abs(filepath.Join(rootPath, "__generated__.yml"))
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to get absolute path: %w", err)
+	}
+
+	generated, err := yamlloader.LoadYAML(virtualPath, bytes.NewReader([]byte(stdout)))
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to parse Python mutator output: %w", err)
+	}
+
+	normalized, diagnostic := convert.Normalize(config.Root{}, generated)
+
+	if diagnostic.Error() != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to normalize Python mutator output: %w", diagnostic.Error())
+	}
+
+	return normalized, nil
+}
+func createOverrideVisitor(ctx context.Context, phase phase) merge.OverrideVisitor {
+	jobsPath := dyn.NewPath(dyn.Key("resources"), dyn.Key("jobs"))
+
+	return merge.OverrideVisitor{
+		VisitDelete: func(valuePath dyn.Path, left dyn.Value) error {
+			if !valuePath.HasPrefix(jobsPath) {
+				return fmt.Errorf("unexpected change at '%s' (delete)", valuePath.String())
+			}
+
+			deleteResource := len(valuePath) == len(jobsPath)+1
+
+			if phase == ApplyPythonMutatorPhasePreInit {
+				return fmt.Errorf("unexpected change at '%s' (delete)", valuePath.String())
+			} else if deleteResource && phase == ApplyPythonMutatorPhaseInit {
+				return fmt.Errorf("unexpected change at '%s' (delete)", valuePath.String())
+			}
+
+			log.Debugf(ctx, "Delete value at '%s'", valuePath.String())
+
+			return nil
+		},
+		VisitInsert: func(valuePath dyn.Path, right dyn.Value) (dyn.Value, error) {
+			if !valuePath.HasPrefix(jobsPath) {
+				return dyn.InvalidValue, fmt.Errorf("unexpected change at '%s' (insert)", valuePath.String())
+			}
+
+			insertResource := len(valuePath) == len(jobsPath)+1
+
+			if phase == ApplyPythonMutatorPhasePreInit && !insertResource {
```

Comment:
Please add a comment to say what's going on here. I think this means it is only possible to insert jobs during `ApplyPythonMutatorPhasePreInit`.